### PR TITLE
[fix]: [DEL-2994]: Cache secrets for expressions

### PR DESCRIPTION
**Pull request by gaurav-jain1**
…ession evaluation

You can use the following comments to re-trigger PR Checks

- Compile: `trigger compile`
- runAeriformCheck: `trigger AeriformCheck`
- CodeFormat: `trigger codeformat`
- MessageMetadata: `trigger messagecheck`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- runDockerizationCheck: `trigger dockerizationcheck`
- runAuthorCheck: `trigger authorcheck`
- Checkstyle: `trigger checkstyle`
- PMD: `trigger pmd`
- TI-bootstrap: `trigger ti0`
- TI-bootstrap1: `trigger ti1`
- TI-bootstrap2: `trigger ti2`
- TI-bootstrap3: `trigger ti3`
- TI-bootstrap4: `trigger ti4`
- FunctionalTest1: `trigger ft1`
- FunctionalTest2: `trigger ft2`
- CodeBaseHash: `trigger codebasehash`

---
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/30190)

### CODE REVIEW

File: 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java [hunk 1]
```diff
@@ -268,6 +273,12 @@
   private Supplier<Long> taskCountCache = Suppliers.memoizeWithExpiration(this::fetchTaskCount, 1, TimeUnit.MINUTES);
   @Inject @Getter private Subject<DelegateTaskStatusObserver> delegateTaskStatusObserverSubject;
 
+  private Cache<String, EncryptedRecordData> secretsCache =
+      Caffeine.newBuilder()
+          .expireAfterWrite(SECRET_CACHE_TTL_MINUTES, TimeUnit.MINUTES)
+          .maximumSize(SECRET_CACHE_MAXIMUM_SIZE)
+          .build();
```
Comment (abhinav-harness) [inline 400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java:280]: we should use distributed cache as pod cache won't reduce the cost by much.

**Comment by jennyjames21** [comment]
___
*[400-rest/src/main/java/software/wings/expression/SecretManagerFunctor.java, line 175 at r1](https://reviewable.io/reviews/harness/harness-core/30190#-MuDbzdf5bUNtl1O5vMd:-MuDbzdf5bUNtl1O5vMe:b-6j061n) ([raw file](https://github.com/harness/harness-core/blob/086ae5a6fdac155327fc0a21e695b74adbe72dfb/400-rest/src/main/java/software/wings/expression/SecretManagerFunctor.java#L175)):*
> ```Java
>         managerDecryptionService.decrypt(serviceVariable, localEncryptedDetails);
>         String randomEncryptionKey = generateUuid();
>         char[] reEncryptedValue = new SimpleEncryption(randomEncryptionKey).encryptChars(serviceVariable.getValue());
> ```

where is serviceVariable.getValue() coming from?

**Comment by jennyjames21** [comment]
___
*[400-rest/src/main/java/software/wings/expression/SecretManagerFunctor.java, line 175 at r1](https://reviewable.io/reviews/harness/harness-core/30190#-MuDbzdf5bUNtl1O5vMd:-MuDc_aPEzukBJ2P_XxY:bea60ja) ([raw file](https://github.com/harness/harness-core/blob/086ae5a6fdac155327fc0a21e695b74adbe72dfb/400-rest/src/main/java/software/wings/expression/SecretManagerFunctor.java#L175)):*
<details><summary><i>Previously, jennyjames21 (Jenny James) wrote…</i></summary><blockquote>

where is serviceVariable.getValue() coming from?  
</blockquote></details>

Wondering we need any error handling here?

**Review by gaurav-jain1 — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30190)* status: 3 file(s) to review. 2. unresolved discussions. 0 of 1 approvals obtained. 

___
*[400-rest/src/main/java/software/wings/expression/SecretManagerFunctor.java, line 175 at r1](https://reviewable.io/reviews/harness/harness-core/30190#-MuDbzdf5bUNtl1O5vMd:-MuEVo5X0PMCSsd0KQTw:bn0na51) ([raw file](https://github.com/harness/harness-core/blob/086ae5a6fdac155327fc0a21e695b74adbe72dfb/400-rest/src/main/java/software/wings/expression/SecretManagerFunctor.java#L175)):*
<details><summary><i>Previously, jennyjames21 (Jenny James) wrote…</i></summary><blockquote>

Wondering we need any error handling here?
</blockquote></details>

That's the decrypted value we get from vault. We used to use the same variable previously without any safety guards. Let me know the specific checks you want me to add.
___
*[400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java, line 280 at r1](https://reviewable.io/reviews/harness/harness-core/30190#-MuA8VpH8upPm1sc5VYa-r1-280:-MuEWBgjDB_BCFJqhnlE:b2gw24d) ([raw file](https://github.com/harness/harness-core/blob/086ae5a6fdac155327fc0a21e695b74adbe72dfb/400-rest/src/main/java/software/wings/service/impl/DelegateTaskServiceClassicImpl.java#L280)):*
<details><summary><i>Previously, abhinav-harness (Abhinav Singh) wrote…</i></summary><blockquote>

we should use distributed cache as pod cache won't reduce the cost by much.
</blockquote></details>

@raghuatwings any thoughts on this ? I am fine with either approach.

**Review by jennyjames21 — APPROVED** [review]
> Reviewed 3 of 3 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30190)* status: 1. unresolved discussions.

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 4 of 4 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30190)* status: :shipit: complete! ✓

**Review by abhinav-harness — APPROVED** [review]
> Reviewed 4 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/30190)* status: :shipit: complete! ✓

**Pull request merged**